rts/motoko-rts/src/stabilization/layout.rs

```diff
@@ -0,0 +1,375 @@
+//! The long-term layout definition of the stable format.
+//!
+//! The stable object graph resides a linear stable memory space.
+//!
+//! Pointers are serialized as skewed offsets in that space.
+//! Scalar and pointer values are serialized as 64-bit `StableValue`
+//! for a long-term perspective, even if main memory operates on 32-bit.
+//! A 32-bit program stores `StableValue` that fit into 32-bit, and
+//! scalar values beyond this width are boxed.
+//! This means that a 32-bit and 64-bit program can always upgrade from
+//! a 32-bit version, while the downgrade from 64-bit to a 32-bit program
+//! is not supported and detected.
+//!
+//! Each object uses a `StableTag` as header an is followed by
+//! the object payload as outlined in the corresponding Rust structs.
+//! Some objects, such as `StableArray`, `StableObject`, `StableBlob`,
+//! and `StableBigNum`
+//!
+//! Not all heap memory object types are stabilized because some
+//! of them are not stable types. New object types can be added
```

Comment:
So are they stabilized to a dummy value?
Say we have a statically stable record that (dynamically) contains a (non)-stable field of function type? That field needs to either be (somehow) elided or (I would expect) serialized to some dummy value. My gut feeling is that would be ok because the field would not be statically reachable so could never be used.

Reply:
My thought was that since stabilization only focuses on the dynamically reachable objects from stable actor variables, those objects would all be of stable types due to the type system. I assume that stable records cannot contain declarations of flexible (non-stable) records because this would indeed complicate things (especially also for enhanced orthogonal persistence). But maybe I miss an aspect?

Reply:
*Update*: I fixed this issue by placing a dummy value in those non-stable fields. Thanks again for pointing out this bug. Still, the deletion of unused fields in new program version is open aspect.

Reply:
I finally got a round to taking a look at the fix, but I think just fixing record fields is not enough, and  the change should actually occur elsewhere, when scanning (?) a closure object.

To see why, I think these are also cases the must be considered:

```
func f() {};
stable let tuple : (Int, Any) = (0, f);
stable let variant : {#tag : Any} = #tag f;
stable let record : { lab : Any} = {lab = f};
stable let vector : [Any] = [f];
stable let array : [var Any] = [var f];
stable let opt : ?Any = ? f;
//etc?
```
Basically, any structured data, even of stable type, could (dynamically) contain a closure, by virtue of subtyping.
So I think the full solution is not  to special case records, but to serialize any TAG_CLOSURE value to some dummy value (perhaps just a scalar is good enough).




stable let

Reply:
Thanks a lot. Indeed, my implementation was wrong and only considered structural sub-typing. I fixed this.